DataProviderService/DataProviderService.cs

```diff
@@ -160,19 +160,19 @@ public static List<Station> StationsFromEDDP(string systemName, dynamic json)
             return Stations;
         }
 
-        public static List<CommodityMarketQuote> CommodityQuotesFromEDDP(dynamic json)
+        public static List<CommodityMarketQuote> CommodityQuotesFromEDDP(JObject json)
         {
             var quotes = new List<CommodityMarketQuote>();
             if (json["commodities"] != null)
             {
-                foreach (dynamic commodity in json["commodities"])
+                foreach (JObject commodity in json["commodities"])
                 {
                     CommodityDefinition commodityDefinition = CommodityDefinition.FromName((string)commodity["name"]);
                     CommodityMarketQuote quote = new CommodityMarketQuote(commodityDefinition);
-                    quote.buyprice = (int?)(long?)commodity["buy_price"];
-                    quote.sellprice = (int?)(long?)commodity["sell_price"];
-                    quote.demand = (int?)(long?)commodity["demand"];
-                    quote.stock = (int?)(long?)commodity["supply"];
+                    quote.buyprice = (int?)(long?)commodity["buy_price"] ?? quote.buyprice;
+                    quote.sellprice = (int?)(long?)commodity["sell_price"] ?? quote.sellprice;
+                    quote.demand = (int?)(long?)commodity["demand"] ?? quote.demand;
+                    quote.stock = (int?)(long?)commodity["supply"] ?? quote.stock;
```

Comment:
Do we need to double cast to nullable values here?

Reply:
Annoyingly I think we do. Unless I have misunderstood, casting directly to `int?` returns null even if a value is present, because the boxed type is `long` not `int`. At any rate, when we first began I removed the double cast and everything broke.

We could call JObject methods like `TryGetValue()` but that would be even uglier.